---
title: "List of Functions"
author: "EJM"
date: "3/13/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

GWAS Function: takes the phenotype, genotype, and covariate data, calculates PCA, calculates colinearity of PC's w/ covariates, constructs GLM equation, applys GLM equation over every SNP, reutrns p-values for association tests

Values
  n = number of individuals (samples)
  m = number of markers
  t = number of covariates

List of Functions
- Function that checks inputs for correct class and dimensions
  - inputs include phenotype (nx1), genotype (nxm), and covariate data (nxt)
  - class of all objects probably needs to be matrix or data frame
  - allows function to proceed or halts function and stops error code
  
- Function that performs PCA on genotype data
  - are there other cofactors we can calculate from genotype data that could increase our accuracy?
  - do we want to use existing PCA functions or write our own?
  - returns PC in ?x? (probably nx1)
  
- Function that calculates colinearity of PC's with covariates
  - calculates colinearity of x number of PCs against x number of covariates
  - what existing function do we want to use to create this matrix?
  - opportunity for extra user input, some sort of colinearity threshold (max correlation value)
  - returns filtered PCs
  
- Function that constructs the GLM equation with all covariates
  - y = SNP + b1X1 ... bnXn + e
  - another opportunity to build in a check point that double checks the equation for fatal errors?
  
- Function that applyes GLM equation over all SNPs
  - returns p-value matrix (1xm)
  
Other Thoughts
- do we want to include the option for the equation to automatically generate a GWAS plot? Fancy!
- Where are residuals calculated?